src/effects/effectprocessor.h

```diff
@@ -6,100 +7,208 @@
+            pState = createState(bufferParameters);
+            m_channelStateMatrix[inputHandle][outputHandle] = pState;
         }
-        return holder.state;
+        processChannel(inputHandle, pState, pInput, pOutput, bufferParameters,
+                       enableState, groupFeatures);
     }
 
-    ChannelHandleMap<ChannelStateHolder> m_channelState;
+   void initialize(const QSet<ChannelHandleAndGroup>& activeInputChannels,
+            EffectsManager* pEffectsManager,
+            const mixxx::AudioParameters& bufferParameters) final {
+        for (const ChannelHandleAndGroup& inputChannel : activeInputChannels) {
+            //qDebug() << this << "EffectProcessorImpl::initialize allocating EffectStates for input" << inputChannel;
+            ChannelHandleMap<EffectSpecificState*> outputChannelMap;
+            for (const ChannelHandleAndGroup& outputChannel :
+                    pEffectsManager->registeredOutputChannels()) {
+                outputChannelMap.insert(outputChannel.handle(),
+                        createState(bufferParameters));
+                //qDebug() << this << "EffectProcessorImpl::initialize registering output" << outputChannel;
+            }
+            m_channelStateMatrix.insert(inputChannel.handle(), outputChannelMap);
+        }
+        m_pEffectsManager = pEffectsManager;
+        DEBUG_ASSERT(m_pEffectsManager != nullptr);
+    };
+
+    EffectSpecificState* createState(const mixxx::AudioParameters& bufferParameters) final {
+        return new EffectSpecificState(bufferParameters);
+    };
+
+    bool loadStatesForInputChannel(const ChannelHandle& inputChannel,
+              const EffectStatesPointer pStates) final {
+          // Can't directly cast a ChannelHandleMap from containing the base
+          // EffectState* type to EffectSpecificState* type, so iterate through
+          // the ChannelHandleMap to build a new ChannelHandleMap with
+          // dynamic_cast'ed states.
+          ChannelHandleMap<EffectSpecificState*> newMap;
+          //qDebug() << "EffectProcessorImpl::loadStatesForInputChannel" << this << "input"
+          //         << inputChannel;
+          for (const ChannelHandleAndGroup& outputChannel :
+                  m_pEffectsManager->registeredOutputChannels()) {
+              //qDebug() << "EffectProcessorImpl::loadStatesForInputChannel" << this << "output" << outputChannel;
+              auto pState = dynamic_cast<EffectSpecificState*>(
+                        pStates->at(outputChannel.handle()));
+              VERIFY_OR_DEBUG_ASSERT(pState != nullptr) {
+                    return false;
+              }
+              newMap.insert(outputChannel.handle(), pState);
+          }
+          m_channelStateMatrix.insert(inputChannel, newMap);
+          return true;
+    };
+
+    // Called from main thread for garbage collection after an input channel is disabled
+    void deleteStatesForInputChannel(const ChannelHandle& inputChannel) final {
```

Comment:
We can't call this from the main thread, because we have concurrent access m_channelStateMatrix then. 
We have to pass the hash for the disabled channel back to the GUI thread.

Reply:
We can call it from the main thread because it is called after the last call to `process` with EffectEnableState::Disabling

Reply:
Yes, but the other channels can still be enabled, so we must not touch the outer hash.

Reply:
I think you are right, but I do not fully understand what could go wrong with this current implementation.

Reply:
Ah yes, you are right. I thought ChannelHandleMap inherits QMap, but this is wrong it inherits QVarLengthArray, which never shrinks. So it would be actually save to remove the Map in case the channel is disabled.  Please rename ChannelHandleMap to ChannelHandleArray to avoid future confusing. It would also be nice to have kMaxExpectedGroups as template parameter. 
This clarifies the situation even more and helps to adjust the preallocated memory to more reasonable values. 

A remaining race condition might be still there if you fast disable and enable a channel. Is it possible to remove a new state instead of the old state? Will removing the old states from the engine thread help here?

Reply:
I don't think that race condition is possible. The call to deleteStatesForInputChannel is always made before writing the request to enable the channel to the MessagePipe.

Reply:
Am I correct about that? Please let me know if I'm overlooking something.

Reply:
I have tried to wrap my brain around it, and I cannot find a race right now.

Reply:
I understand your confusion regarding ChannelHandleMap not inheriting from QMap, but I would be confused by renaming it to ChannelHandleArray. An array is typically a sequential list indexed only by integer (unless you're writing PHP). ChannelHandleMaps are indexed by ChannelHandles, so I think the name makes sense.

Reply:
It is actually that, an handle is an integer. The word map is simply wrong here.

Reply:
A ChannelHandle is not an integer. Its internal use of an integer is a detail that has no meaning outside of its implementation.

Reply:
We rely on the array nature here, so it is an important details. If we ignore the array nature, we need to pass the memory via the message because we have to assume a map.
A map is a sorted array with fast lookups, but less memory. An array as instant lookups. This is an array.

Reply:
The new table prefix will solve this here.

Reply:
Ping!